programs/bpf_loader/src/syscalls.rs

```diff
@@ -2041,6 +2062,107 @@ impl<'a, 'b> SyscallObject<BpfError> for SyscallZkTokenElgamalOpWithScalar<'a, '
     }
 }
 
+// Elliptic curve point validation
+pub struct SyscallCurveValidatePoint<'a, 'b> {
+    invoke_context: Rc<RefCell<&'a mut InvokeContext<'b>>>,
+    check_aligned: bool,
+}
+
+impl<'a, 'b> SyscallObject<BpfError> for SyscallCurveValidatePoint<'a, 'b> {
+    fn call(
+        &mut self,
+        curve_id: u64,
+        left_addr: u64,
+        result_addr: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        use {
+            num_traits::FromPrimitive,
+            solana_zk_token_sdk::curve25519::{
+                curve_syscall_traits::CurveId,
+                edwards::{validate_edwards, PodEdwardsPoint},
+                ristretto::{validate_ristretto, PodRistrettoPoint},
+            },
+        };
+
+        let invoke_context = question_mark!(
+            self.invoke_context
+                .try_borrow()
+                .map_err(|_| SyscallError::InvokeContextBorrowFailed),
+            result
+        );
+
+        if let Some(curve_id) = FromPrimitive::from_u64(curve_id) {
+            match curve_id {
+                CurveId::Curve25519Edwards => {
```

Comment:
The code for these two branches is different only in choice between Edwards/Ristretto. Maybe it is possible to make a function or a macro to simplify that?

Reply:
The curve implementations for the two curves are completely different, so I am not sure if it would be possible. But worth thinking about...